client/components/data/query-sites/index.jsx

```diff
@@ -3,19 +3,31 @@
  */
 import { Component, PropTypes } from 'react';
 import { connect } from 'react-redux';
-import { bindActionCreators } from 'redux';
 
 /**
  * Internal dependencies
  */
-import { isRequestingSites } from 'state/sites/selectors';
-import { requestSites } from 'state/sites/actions';
+import { isRequestingSites, isRequestingSite } from 'state/sites/selectors';
+import { requestSites, requestSite } from 'state/sites/actions';
 
 class QuerySites extends Component {
-
 	componentWillMount() {
-		if ( ! this.props.requestingSites ) {
-			this.props.requestSites();
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.siteId !== this.props.siteId ) {
+			this.request( nextProps );
+		}
+	}
+
+	request( props ) {
+		if ( ! props.siteId && ! props.requestingSites ) {
```

Comment:
This shouldn't hold up the PR, but I think splitting it into two keeps it simple and makes it easier to use. 

Passing an explicit `null` parameter works too, but we'd need to communicate the convention. Currently, selectors often return `null` so this would be easy to accidentally trigger.